### PR TITLE
Completed lecture1, started lecture 2

### PR DESCRIPTION
**Pull request by Mellowdv**
### Solutions for Lecture 1 (and a little of 2)

cc @chshersh

### CODE REVIEW

File: src/Lecture1.hs [hunk 1]
```diff
@@ -30,6 +30,7 @@ module Lecture1
     , strSum
     , lowerAndGreater
     ) where
+import GHC.RTS.Flags (GCFlags(compact))
```
Comment (chshersh) [inline src/Lecture1.hs:33]: I don't think you intended to import this module. I guess, VSCode automatically inserted it for you.

In any case, you can safely remove it 🙂

File: src/Lecture1.hs [hunk 2]
```diff
@@ -67,7 +70,8 @@ sumOfSquares x y = error "TODO!"
 
 -}
 -- DON'T FORGET TO SPECIFY THE TYPE IN HERE
-lastDigit n = error "lastDigit: Not implemented!"
+lastDigit :: Int -> Int
+lastDigit n = mod (abs n) 10
```
Comment (chshersh) [inline src/Lecture1.hs:74]: Absolutely correct 👍🏻 
Alternatively, you can use the infix form of a function by putting it inside backticks :slightly_smiling_face: 
```suggestion
lastDigit n = abs n `mod` 10
```

File: src/Lecture1.hs [hunk 3]
```diff
@@ -108,7 +120,9 @@ and finds a sum of the numbers inside this string.
 
 The string contains only spaces and/or numbers.
 -}
-strSum str = error "TODO"
+strSum :: [Char] -> Int
+strSum str = sum numberList
+    where numberList = map read (words str)
```
Comment (chshersh) [inline src/Lecture1.hs:125]: If you prefer, you can even implement this function as a one-liner

```suggestion
strSum str = sum (map read (words str))
```

File: src/Lecture1.hs [hunk 4]
```diff
@@ -123,4 +137,13 @@ and lower than 6 elements (4, 5, 6, 7, 8 and 9).
...
+lowerAndGreater :: Int -> [Int] -> [Char]
+lowerAndGreater n list = show n ++ " is greater than " ++ show (head counts) ++ " elements and lower than " ++ show (last counts) ++ " elements"
+    where
+        go :: [Int] -> [Int] -> [Int]
+        go result l
+            | null l = result
+            | n > head l = go [succ (head result), last result] (tail l)
+            | n < head l = go [head result, succ (last result)] (tail l)
+            | otherwise = go result (tail l)
+        counts = go [0, 0] list
```
Comment (chshersh) [inline src/Lecture1.hs:149]: You can pass these two numbers as two separate arguments. This way you won't need to call `head` and `last` on every recursive call

File: src/Lecture2.hs [hunk 5]
```diff
@@ -62,7 +65,9 @@ lazyProduct = error "TODO"
 "ccaabb"
 -}
 duplicate :: [a] -> [a]
-duplicate = error "TODO"
+duplicate list = case list of
+     [var] -> [var, var]
+     _ -> head list : head list : duplicate (tail list)
```
Comment (chshersh) [inline src/Lecture2.hs:70]: Try to use `[]` and `(x : xs)` patterns on `list` instead of `[var]` and the `_`

File: src/Lecture2.hs [hunk 6]
```diff
@@ -52,7 +52,10 @@ zero, you can stop calculating product and return 0 immediately.
 84
 -}
 lazyProduct :: [Int] -> Int
-lazyProduct = error "TODO"
+lazyProduct list = case list of
+    (0 : _) -> 0
+    [num] -> num
+    _ -> head list * lazyProduct (tail list)
```
Comment (chshersh) [inline src/Lecture2.hs:58]: Usually, when you call `head` and `tail` on the same list, it's a good chance you want to pattern match on a list instead:

```suggestion
    x : xs -> x * lazyProduct xs
```

**Review by chshersh — APPROVED** [review]
Congrats! The first homework is complete 👏🏼 
Hope you enjoyed the course so far. Keep up the great work, and feel free to ask any questions if you have 👌🏼

File: src/Lecture1.hs (hunk 1 above)
Comment (Mellowdv) [reply src/Lecture1.hs:33]: Yes, I accidentally auto completed something and it got imported.

File: src/Lecture1.hs (hunk 4 above)
Comment (Mellowdv) [reply src/Lecture1.hs:149]: I need to return them in a tuple then, correct?

File: src/Lecture2.hs (hunk 6 above)
Comment (Mellowdv) [reply src/Lecture2.hs:58]: Right, true! I thought `x : xs` wasn't very expressive but I suppose it's clear and understandable to people who write Haskell, I'll use it.

File: src/Lecture2.hs (hunk 5 above)
Comment (Mellowdv) [reply src/Lecture2.hs:70]: Okay, I'll redo this for the lecture2 exercises PR. :)

File: src/Lecture1.hs (hunk 4 above)
Comment (chshersh) [reply src/Lecture1.hs:149]: You have several options here:

* Return a two-element list as you do now but only in the end
* Return a tuple to guarantee that the function returns exactly 2 elements (however, I haven't explained tuples in lecture 1)
* Return `String`: when you reach the end of recursion, you can construct the string as the result. This way you don't need to return numbers.

**Pull request merged**